Update bbr-etcd job to use new bbr backup_one_restore_all feature

aclevername opened the pull request:
### WHAT is this change about?

This change fixes backup and restore of a multi-master cluster when using the latest version of BBR

### WHY is this change being made (What problem is being addressed)?

In the latest [BBR CLI v1.5.0](https://github.com/cloudfoundry-incubator/bosh-backup-and-restore/releases/tag/v1.5.0) we remove support for setting custom `backup_name` and `restore_name` in the `bin/bbr/metadata.sh` script. This means that with the latest version of the BBR CLI user will not currently be able to restore. With this PR we are fixing this by migrating to using a new BBR feature called `backup_one_restore_all`.

### Please provide contextual information.

Story for this PR https://www.pivotaltracker.com/story/show/164144061
Story for the `backup_one_restore_all` feature https://www.pivotaltracker.com/story/show/164144014

### Does this PR introduce a breaking change? 

This PR introduces a new mandatory manifest property

```
properties:
  bbr.backup_one_restore_all:
    default: false
```

This property is required to be present inside the BOSH manifest in order for the BBR CLI to detect it. When BBR reads this property it knows to restore the same artifact to all instances, in this case taking a backup artifact from one of the master nodes and restoring it to all nodes. This is to prevent split brain/inconsistencies across the etcd nodes.

Due to how BBR CLI detects this property it **has to be present in the manifest**, because of this we have setup the default value to be `false` and have added some erb templating to ensure that this fails at templating time when the value is set to `false`. This **enforces the operator to set the value to `true` in the manifest**. 

### Anything else?

We created some rspec tests in order to drive out the templating. Feel free to add them to your CI if you like (they are quite quick).

### Tag your pair, your PM, and/or team!

@alamages @glestaris @MModhaPivotal



Thanks,
Jake & Mehul

alamages committed:
Update bbr-etcd job to use new bbr backup_one_restore_all feature

bbr named artifacts feature, using metadata script, will soon be deprecated.
This commit updates the job to use the new backup_one_restore_all
feature which will replace the named artifacts. Since this is a new
feature the job should fail if the bbr version being run does not
support it.

We used spec tests to drive out the behaviour in the templates.

[#164144061]

Signed-off-by: Mehul Modha <mmodha@pivotal.io>

afong94 commented:
Hi @aclevername, I just tested your changes with k-drats and I am seeing the following issue:

```bbr] 2019/03/19 14:43:34 INFO - Backing up bbr-etcd on master/561c51c9-9548-4592-a806-9aab93d960db...
[bbr] 2019/03/19 14:43:34 INFO - Backing up bbr-etcd on master/daf2f385-abe2-4c55-b15b-ae63c423f102...
[bbr] 2019/03/19 14:43:34 ERROR - Error backing up bbr-etcd on master/561c51c9-9548-4592-a806-9aab93d960db.
[bbr] 2019/03/19 14:43:34 ERROR - Error backing up bbr-etcd on master/daf2f385-abe2-4c55-b15b-ae63c423f102.
[bbr] 2019/03/19 14:43:34 ERROR - Error backing up bbr-etcd on master/40501502-c57b-483a-b585-c9d1f08fab8e.
[bbr] 2019/03/19 14:43:34 INFO - Finished running backup scripts.
[bbr] 2019/03/19 14:43:34 INFO - Running post-backup-unlock scripts...
[bbr] 2019/03/19 14:43:34 INFO - Finished running post-backup-unlock scripts.
1 error occurred:
error 1:
3 errors occurred:
error 1:
Error attempting to run backup for job bbr-etcd on master/561c51c9-9548-4592-a806-9aab93d960db: /var/vcap/jobs/bbr-etcd/bin/bbr/backup: 4: set: Illegal option -o pipefail - exit code 2error 2:
Error attempting to run backup for job bbr-etcd on master/daf2f385-abe2-4c55-b15b-ae63c423f102: /var/vcap/jobs/bbr-etcd/bin/bbr/backup: 4: set: Illegal option -o pipefail - exit code 2error 3:
Error attempting to run backup for job bbr-etcd on master/40501502-c57b-483a-b585-c9d1f08fab8e: /var/vcap/jobs/bbr-etcd/bin/bbr/backup: 4: set: Illegal option -o pipefail - exit code 2


STEP: running bbr deployment backup-cleanup
[bbr] 2019/03/19 14:43:39 INFO - Looking for scripts
[bbr] 2019/03/19 14:43:41 INFO - master/40501502-c57b-483a-b585-c9d1f08fab8e/bbr-etcd/backup
[bbr] 2019/03/19 14:43:41 INFO - master/40501502-c57b-483a-b585-c9d1f08fab8e/bbr-etcd/metadata
[bbr] 2019/03/19 14:43:41 INFO - master/40501502-c57b-483a-b585-c9d1f08fab8e/bbr-etcd```

Was the shell changed to use sh instead as well in the bbr cli?

alamages committed:
Move ruby block in backup script

This was causing the first line to be a space instead of the #! and
therefore file was treated as a sh (and not bash)

[#164144061]

Signed-off-by: Glen Rodgers <grodgers@pivotal.io>

gmrodgers committed:
Metadata script errors are now directed to stderr

[#164144061]

Signed-off-by: Emmanouil Kiagias <ekiagias@pivotal.io>

alamages commented:
Hello @afong94,

We missed this when testing. We have updated the PR. It should now work :). Let us know if you need anything else.

cc @gmrodgers

afong94 commented:
Tested with k-drats on a modified cluster with these changes along with https://github.com/cloudfoundry-incubator/kubo-deployment/pull/383. 

LGTM
